bot/utils/services.py

```diff
@@ -57,5 +113,170 @@ async def send_to_paste_service(contents: str, *, extension: str = "") -> Option
 
         log.warning(
             f"Got unexpected JSON response from paste service: {response_json}\n"
-            f"trying again ({attempt}/{FAILED_REQUEST_ATTEMPTS})."
+            f"trying again ({_attempt}/{_attempts})."
         )
+
+    return await _attempt_request(paste_url, FAILED_REQUEST_ATTEMPTS, data=contents, callback=_callback)
+
+
+async def _get_url_from_cache(url: str) -> Optional[_UnfurlReturn]:
+    """Return an unfurled URL from cache, or None if it doesn't exist or is expired."""
+    cached = await UNFURL_CACHE.get(url)
+
+    if cached is not None:
+        log.trace(f"Found hit for URL ({log}) in unfurl cache.")
+        data = json.loads(cached)
+        expiry = datetime.datetime.fromisoformat(data["expiry"])
+
+        if expiry < datetime.datetime.now(tz=datetime.timezone.utc):
+            # Cache expired, remove it and continue normally
+            log.debug(f"Cache entry for ({url}) expired, deleting.")
+            await UNFURL_CACHE.delete(url)
+        else:
+            # Found unexpired hit
+            return _UnfurlReturn(data["destination"], data["depth"], None, expiry - CACHE_LENGTH)
+
+
+async def _unfurl_url(
```

Comment:
The amount of return types makes this function confusing. It can be greatly simplified by making the request outside this function and then giving this function the response object. Here's my untested attempt:

```py
async def _cache_url(url: str, result: _UnfurlReturn) -> None:
    """TODO."""
    data = {
        "destination": result.destination,
        "depth": result.depth,
        "expiry": (result.created_at + CACHE_LENGTH).isoformat()
    }

    # Write to cache and return results
    log.trace(f"Writing URL ({url}) to unfurl CACHE.")
    await UNFURL_CACHE.set(url, json.dumps(data))


async def _handle_unfurl_response(
    response: aiohttp.ClientResponse,
    url: str,
    redirects: int,
) -> _UnfurlReturn:
    """TODO."""
    # See link for documentation on how to use the worker
    # https://github.com/python-discord/workers/tree/main/url-unfurler
    if response.status == 200:
        # Success, return the destination
        data = await response.json()
        now = datetime.datetime.now(tz=datetime.timezone.utc)
        return _UnfurlReturn(data["destination"], redirects + data["depth"], None, now)

    elif response.status == 400:
        # Unrecoverable error with the URL used for this unfurling
        error_message = (await response.json())["error"]
        log.warning(f"Failed to unfurl URL ({url}) with error message: {error_message}")
        return _UnfurlReturn(error=error_message)

    elif response.status == 416:
        # Max depth reached by the worker, try again or exit
        data = await response.json()
        redirects += data["depth"]
        log.info(f"Failed to unfurl URL ({url}) after {redirects} redirects.")
        return _UnfurlReturn(data["next"], redirects, data["error"])

    elif response.status == 418:
        # The redirect chain has a broken link, can not continue
        data = await response.json()
        redirects += data["depth"]

        log.warning(
            f"Failed to unfurl URL ({url}) due to a broken link in the chain. "
            f"Followed {redirects} redirects, and stopped at ({data['final']})."
        )
        return _UnfurlReturn(data["final"], redirects, data["error"])

    else:
        response.raise_for_status()


async def unfurl_url(
    url: str,
    *,
    max_per_attempt: int = MAX_UNFURLS,
    max_continues: int = 0,
    calculate_continues: bool = False,
    use_cache: bool = True,
) -> Optional[_UnfurlReturn]:
    """
    Follow all redirects of a URL, and return the final address.

    Returns an object containing the final URL, number of redirects, error text, and date of last fetch.
    Returns None if we couldn't resolve the URL for any reason.

    The error in the return is only set if the operation failed.
    If false, the final destination might not be accurate, and the date will be None.

    `max_per_attempt` is passed directly to the worker. If the worker errors out due to too many redirects,
    we'll attempt to restart from the last URL `max_continues` times.

    If `calculate_continues` is true, we'll paginate `max_per_attempt` into as many continues as necessary.
    """
    if not calculate_continues and max_per_attempt > MAX_UNFURLS:
        raise ValueError(f"Max attempts `{max_per_attempt}` is greater than the maximum allowable `{MAX_UNFURLS}`.")

    max_redirects = 0
    if calculate_continues:
        max_continues = (max_per_attempt // MAX_UNFURLS) - (1 if max_per_attempt % MAX_UNFURLS == 0 else 0)
        max_redirects = max_per_attempt

    next_url = url
    redirects = 0
    continues = 0

    if not use_cache:
        log.info(f"Skipping cache for URL unfurling ({url})")

    while True:
        # Check if we can just use the cache
        if use_cache and (hit := await _get_url_from_cache(next_url)) is not None:
            return hit

        if calculate_continues:
            redirects_left = max_redirects - redirects
            max_per_attempt = max(min(redirects_left, MAX_UNFURLS), 0)  # Get a number bound between 0 and MAX_UNFURLS

        body = {"url": next_url, "max-depth": max_per_attempt}
        response = await _attempt_request(URLs.unfurl_worker, 3, json=body, raise_for_status=False)
        if response is None:
            return response

        if response.status == 416 and continues < max_continues:
            # Try again (the caller will make a new call with the next URL)
            data = await response.json()
            log.debug(
                f"Hit maximum depth while unfurling URL ({url}) after {data['depth']} attempts. "
                f"Last request was made to ({data['final']}), and next request will be made to {data['next']}. "
                f"{continues + 1}/{max_continues} failures."
            )

            next_url = data['next']
            redirects += data['depth'] + 1
            continues += 1
        else:
            result = await _handle_unfurl_response(response, next_url, redirects)
            if response.status == 200:
                await _cache_url(url, result)
            return result
```

Reply:
`_unfurl_url_once` is probably a better name if you don't follow my refactoring suggestion.